flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/strategy/AdaptiveSkewedJoinOptimizationStrategy.java

```diff
@@ -0,0 +1,317 @@
+                    isModificationSucceed ? "succeeded" : "failed",
+                    adaptiveJoinNode.getId());
+        }
+        if (isRightOptimizable) {
+            boolean isModificationSucceed =
+                    postStreamEdgeModificationRequest(
+                            context, adaptiveJoinNode, RIGHT_INPUT_TYPE_NUMBER);
+            LOG.info(
+                    "Apply skewed join optimization {} for right input of node {}.",
+                    isModificationSucceed ? "succeeded" : "failed",
+                    adaptiveJoinNode.getId());
+        }
+    }
+
+    private static boolean postStreamEdgeModificationRequest(
+            StreamGraphContext context, ImmutableStreamNode adaptiveJoinNode, int typeNumber) {
+        List<StreamEdgeUpdateRequestInfo> modifiedRequests = new ArrayList<>();
+        modifiedRequests.addAll(
+                generateModifiedCorrelationRequestInfos(
+                        filterEdges(adaptiveJoinNode.getInEdges(), typeNumber)));
+        modifiedRequests.addAll(
+                generateForwardPartitionerModificationRequestInfos(
+                        adaptiveJoinNode.getOutEdges(), context));
+        return context.modifyStreamEdge(modifiedRequests);
+    }
+
+    private static List<StreamEdgeUpdateRequestInfo> generateModifiedCorrelationRequestInfos(
+            List<ImmutableStreamEdge> streamEdges) {
+        List<StreamEdgeUpdateRequestInfo> streamEdgeUpdateRequestInfos = new ArrayList<>();
+        for (ImmutableStreamEdge edge : streamEdges) {
+            streamEdgeUpdateRequestInfos.add(
+                    new StreamEdgeUpdateRequestInfo(
+                                    edge.getEdgeId(), edge.getSourceId(), edge.getTargetId())
+                            .existIntraInputKeyCorrelation(false));
+        }
+        return streamEdgeUpdateRequestInfos;
+    }
+
+    private static List<StreamEdgeUpdateRequestInfo>
+            generateForwardPartitionerModificationRequestInfos(
+                    List<ImmutableStreamEdge> streamEdges, StreamGraphContext context) {
+        List<ImmutableStreamEdge> forwardEdges =
+                streamEdges.stream()
+                        .filter(ImmutableStreamEdge::isForwardEdge)
+                        .collect(Collectors.toList());
+        List<StreamEdgeUpdateRequestInfo> streamEdgeUpdateRequestInfos = new ArrayList<>();
+        for (ImmutableStreamEdge edge : forwardEdges) {
+            StreamPartitioner<?> newPartitioner;
+            StreamPartitioner<?> partitioner =
+                    context.getOutputPartitioner(
+                            edge.getSourceId(), edge.getTargetId(), edge.getEdgeId());
+            if (partitioner instanceof ForwardForConsecutiveHashPartitioner) {
+                newPartitioner =
+                        ((ForwardForConsecutiveHashPartitioner<?>) partitioner)
+                                .getHashPartitioner();
+            } else {
```

Comment:
Only `ForwardForUnspecifiedPartitioner` can be converted to `RescalePartitioner`.